### PR TITLE
feat(device): introduce web of things support

### PR DESCRIPTION
**Pull request by paga16-hash**
In this PR the frontend and the device microservice has been adapted to open the software to the Web of Things. 
[Node WoT](https://github.com/eclipse-thingweb/node-wot) has been used and a sample thing can be found in the organization repositories.
Let me know if some changes are needed 🗡️

### CODE REVIEW

File: device/src/application/services/impl/DeviceServiceImpl.ts [hunk 1]
```diff
@@ -6,49 +6,62 @@ import { Device } from '@/domain/core/Device.js'
...
-                  res.data.capabilities[i].measure.type,
-                  res.data.capabilities[i].measure.unit
+    return await this.servient
+      .start()
+      .then(async (WoT: any): Promise<DeviceCapability[]> => {
+        const td = await WoT.requestThingDescription(
+          `http://${device.endpoint.ipAddress}:${device.endpoint.port}/device`
+        )
+        const thing = await WoT.consume(td)
+        return thing.readProperty('status').then(async (data: any): Promise<DeviceCapability[]> => {
```
Comment (Mala1180) [inline device/src/application/services/impl/DeviceServiceImpl.ts:40]: Wonderful

**Comment by paga16-hash** [comment]
> Hey @paga16-hash, unfortunately this PRs has some conflicts... 😞 Could you please fix them? Thks 😘

File: device/src/application/services/impl/DeviceServiceImpl.ts (hunk 1 above)
Comment (paga16-hash) [reply device/src/application/services/impl/DeviceServiceImpl.ts:40]: Thanks :D

Comment (paga16-hash) [reply device/src/application/services/impl/DeviceServiceImpl.ts:40]: Take a look to the [Thing](https://github.com/revue-org/revue-sample-thing)

Comment (kelvin-olaiya) [reply device/src/application/services/impl/DeviceServiceImpl.ts:40]: Nice!
I can't see the ```Wot``` object import... I believe is missing

File: device/src/application/services/impl/DeviceServiceImpl.ts [hunk 2]
```diff
@@ -74,29 +87,27 @@ export class DeviceServiceImpl implements DeviceService {
...
-            (capability: any): CapabilityType => {
-              if (Object.values(CapabilityType).includes(capability.type as CapabilityType)) {
+      await this.servient.start().then(async (WoT: any): Promise<void> => {
+        const td = await WoT.requestThingDescription(
+          `http://${device.endpoint.ipAddress}:${device.endpoint.port}/device`
+        )
+        const thing = await WoT.consume(td)
+        await thing.readProperty('status').then(async (data: any): Promise<void> => {
+          const deviceStatus = await data.value()
+          if (deviceStatus.enabled) {
```
Comment (kelvin-olaiya) [inline device/src/application/services/impl/DeviceServiceImpl.ts:99]: Why are we filtering out disabled devices?

File: device/src/application/services/impl/DeviceServiceImpl.ts [hunk 3]
```diff
@@ -105,25 +116,36 @@ export class DeviceServiceImpl implements DeviceService {
...
-    )
-    await this.repository.saveDevice(device)
-    this.events.publishDeviceAdded(DeviceEventFactory.createAddition(new Date(), device.deviceId.value))
-    return device.deviceId
+  async createDevice(description: string, endpoint: DeviceEndpoint): Promise<DeviceId> {
+    return await this.servient.start().then(async (WoT: any): Promise<DeviceId> => {
+      const td = await WoT.requestThingDescription(`http://${endpoint.ipAddress}:${endpoint.port}/device`)
+      const thing = await WoT.consume(td)
+      return await thing.readProperty('status').then(async (data: any): Promise<DeviceId> => {
+        const deviceStatus = await data.value()
```
Comment (kelvin-olaiya) [inline device/src/application/services/impl/DeviceServiceImpl.ts:124]: @paga16-hash Yes, a "Thing manager" as you mention before would be nice here and there. (A nice to have, let's say)

**Review by kelvin-olaiya — COMMENTED** [review]
Overall I believe it is well done. I left just few comments.

File: device/src/application/services/impl/DeviceServiceImpl.ts (hunk 1 above)
Comment (paga16-hash) [reply device/src/application/services/impl/DeviceServiceImpl.ts:40]: these are the imports: one for the Servient and one for the HttpClientServer
import { Servient } from '@node-wot/core'
import HttpClientFactory from '@node-wot/binding-http'
The WoT object is created by the Servient after its start.

If you want you can also take a look to the example in the WoT doc [here](https://github.com/eclipse-thingweb/node-wot#examples). Let me know if i clarified your dubt :D

File: device/src/application/services/DeviceService.ts [hunk 4]
```diff
@@ -11,13 +11,7 @@ export interface DeviceService {
 
   getDeviceById(deviceId: DeviceId): Promise<Device>
 
-  createDevice(
-    description: string,
-    endpoint: DeviceEndpoint,
-    locationId: string,
```
[the remark this answers is not in the document]
Comment (paga16-hash) [reply device/src/application/services/DeviceService.ts:17]: yes, we can think to an action on the Thing to update it, but i don't think is too correct. The location depends on the device.

[the remark this answers is not in the document]
Comment (kelvin-olaiya) [reply device/src/application/services/DeviceService.ts:17]: Ok, seems reasonable. The only problem i see is that I believe that the device must be able to contact revue to get the list of existing environments from which to choose. This means Authentication (which another interaction device-revue, and not the other way round). Do you agree?

File: device/src/application/services/impl/DeviceServiceImpl.ts (hunk 2 above)
Comment (kelvin-olaiya) [reply device/src/application/services/impl/DeviceServiceImpl.ts]: 🔝

File: device/src/application/services/impl/DeviceServiceImpl.ts (hunk 3 above)
Comment (kelvin-olaiya) [reply device/src/application/services/impl/DeviceServiceImpl.ts:122]: As you prefer. You can take the opportunity to do some TDD if you want. Anyways, you can plan it for a next PR without any problem in my opinion. @Mala1180 what do you think?

Comment (paga16-hash) [reply device/src/application/services/impl/DeviceServiceImpl.ts:122]: I think its better to do it in a next PR so i can proceed with general tests and then refactor this part of code. Same for device tests, we have to mock a thing or remove the post test..

Comment (Mala1180) [reply device/src/application/services/impl/DeviceServiceImpl.ts:122]: Next PR is ok

**Review by Mala1180 — APPROVED** [review]

**Review by kelvin-olaiya — DISMISSED** [review]

**Review by Mala1180 — APPROVED** [review]

**Review by Mala1180 — DISMISSED** [review]

**Review by kelvin-olaiya — APPROVED** [review]

**Review by kelvin-olaiya — APPROVED** [review]

**Review by Mala1180 — APPROVED** [review]

**Pull request merged**